### PR TITLE
ck_assert_str_*: fix segmentation fault while comparing to NULL

### PR DESCRIPTION
**Pull request by 9fcc**
Often a string need to be compared to NULL in tests. And in some cases compared string could be NULL so it must not lead to segmentation fault. Comparing NULL to NULL with *_eq macro must be true. In any other situations comparing to NULL must be false. *_le and *_ge macros must return false because NULL could not be lesser or greater than string. So returning true in *_le or *_ge with both arguments passed as NULL has no sense.

### CODE REVIEW

**Comment by 9fcc** [comment]
My colleague used those macros to check a string pointer that can have NULL value. NULL value is used as separate state to show that string was not set and must be ignored in further code. We often use NULL value to add some different state of variable so it must be checkable too.

**Comment by brarcher** [comment]
I'm a little unsure about this, as the string macros would be expected to operate on valid strings, and C11 §7.1.1 defines a string as a sequence of characters with a terminating \0 character. That is, the pointer itself must be valid.

If one is interested in checking if a string is non-null, there are some ck_assert_ptr_* macros that one could use. For example, one could:
```
ck_assert_ptr_eq(myptr, NULL);
```
One could also contribute other macros, if they are useful, such as ck_assert_ptr_null/nonnull. These currently do not exist.

What may be useful to avoid a segfault is to add additional checks to ck_assert_str_* to verify that the pointers are non-null and fail if they are. That would be in line with the spirit of C11 and its definition of strings.

**Comment by 9fcc** [comment]
I disagree. You're talking about strings but not a pointer to a string. I don't think that only our company uses NULL state of string pointer. It's very convenient to use additional state.

Let's see an example. We have a username variable. If it is set by some function we run some process with this appropriate privileges. It it is not set we run process under the same privileges as the parent process. Now want to test this feature. We call the function and set username. And then we check then username is correct. But if we didn't set username we want to check if username is NULL. Using prt_eq is this case is nonsense. It's a string for us that is just not set.

One more example. I want to make test universal. I run cycle passing different users to the test to check against an array item. If I would user ptr_eq I'll need to add one more 'if' condition. Just more code.

For now tests just crash when comparing to NULL. With my changes  all would be the same as in current version but with no crashes. And moreover macros would be more useful as they flexible.

API must provide all the user needs.

If you still doubt, I can add a define that would allow such a comparison. In some years it would show how much users uses that macro. If that number would be significant it could become default behaviour..

**Comment by 9fcc** [comment]
> What may be useful to avoid a segfault is to add additional checks to ck_assert_str_* to verify that the pointers are non-null and fail if they are. That would be in line with the spirit of C11 and its definition of strings.

We already use if condition and ptr_eq macro in current version. It just increased the code amount. And nothing else.

**Comment by brarcher** [comment]
Jenkins: ok to test

**Comment by brarcher** [comment]
Your argument is convincing. Let's move forward with this change to allow the state for an undefined (NULL) string for the eq and ne str macros.

File: src/check.h.in [hunk 1]
```diff
@@ -690,7 +699,7 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
  *
  * @since 0.9.6
  */
```
Comment (brarcher) [inline src/check.h.in]: There may need to be additional documentation on these macros, stating that NULL is allowed for *_eq and *_ne, but if a pointer is NULL for any other macro the macro will report a test failure.

File: tests/check_check_sub.c [hunk 2]
```diff
@@ -553,6 +555,7 @@ START_TEST(test_ck_assert_str_le)
   record_test_name(tcase_name());
   ck_assert_str_le(s, t);
   ck_assert_str_le(s, s);
+  ck_assert_str_le(NULL, NULL);
```
Comment (brarcher) [inline tests/check_check_sub.c]: Oh, I see, this is resolved in a follow-on commit. Could the commits be combined?

**Comment by brarcher** [comment]
I think that the documentation will also need to be updated to mention that NULL is now allowed for the string macros.

**Comment by 9fcc** [comment]
I've updated documentation and macros doxygen description.

File: src/check.h.in [hunk 3]
```diff
@@ -670,18 +670,30 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
...
+    #X" "#OP" "#Y, \
+    #X, _ck_x_q, _ck_x, _ck_x_q, \
+    #Y, _ck_y_q, _ck_y, _ck_y_q); \
 } while (0)
+
 /**
  * Check two strings to determine if 0==strcmp(X,Y)
  *
- * If not 0==strcmp(X,Y), the test fails.
+ * If ((X==NULL)||(Y==NULL))&&(X!=Y), the test fails.
```
Comment (brarcher) [inline src/check.h.in]: Although this is correct, it may be a bit to digest. Would the following be acceptable:
```
If both X and Y are NULL the test passed. However, if only one is NULL the test fails.
```

File: src/check.h.in [hunk 4]
```diff
@@ -690,11 +702,13 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
...
  * @since 0.9.6
  */
-#define ck_assert_str_eq(X, Y) _ck_assert_str(X, ==, Y)
+#define ck_assert_str_eq(X, Y) _ck_assert_str(X, ==, Y, 1, 0)
+
 /**
  * Check two strings to determine if 0!=strcmp(X,Y)
  *
- * If not 0!=strcmp(X,Y), the test fails.
+ * If (X==NULL)&&(Y==NULL), the test fails.
```
Comment (brarcher) [inline src/check.h.in]: The same applies here.
```
If either X or Y is NULL the test passes, however if both are NULL the test fails.
```

File: src/check.h.in [hunk 5]
```diff
@@ -703,10 +717,12 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
  *
  * @since 0.9.6
  */
-#define ck_assert_str_ne(X, Y) _ck_assert_str(X, !=, Y)
+#define ck_assert_str_ne(X, Y) _ck_assert_str(X, !=, Y, 0, 1)
+
 /**
  * Check two strings to determine if 0<strcmp(X,Y), (e.g. strcmp(X,Y)>0)
  *
+ * If (X==NULL)||(Y==NULL), the test fails.
```
Comment (brarcher) [inline src/check.h.in]: This could be simplified here (and for the remaining macros):
```
If X or Y is NULL the test fails.
```

**Comment by brarcher** [comment]
Besides the minor comments to the documentation, this is good to go.

File: src/check.h.in (hunk 3 above)
Comment (9fcc) [inline src/check.h.in]: > the test passed.

Here must be `passes`, am I right?

**Comment by 9fcc** [comment]
Done.

**Comment by chrisnc** [comment]
This would be very surprising behavior to me. The fact that the str and mem comparison macros map straightforwardly to `strcmp` and `memcmp` makes them conceptually simple; those who understand `strcmp` and `memcmp` can easily grasp how this API works. Also, there are cases where two strings being null in a comparison that is intended to compare string data would silently pass an assertion with this change. The workaround for users who use `NULL` as a valid state of a variable is very simple: an additional assertion. Don't opt in all users to this behavior.

One alternative that seems reasonable and consistent to me is to have the str and mem comparison macros also assert that their arguments are non-NULL before attempting to compare them. This prevents crashes and also enforces the invariants of the API (which are identical to that of `strcmp` and `memcmp`, as they should be).

This blog post by Colin Percival (of Tarsnap and FreeBSD) pertains to a related problem in `printf` which can cause silent bugs by accepting `NULL` for string pointers and silently print `"(null)"` instead of aborting.

http://www.daemonology.net/blog/2015-08-13-null-is-a-bad-idea.html

**Comment by 9fcc** [comment]
It's tests. You know what values could be passed to your function. So You can add assert to check for NULL before comparing strings if it is erroneous for you. Some libraries already use this feature (but not exactly the same way).
[https://developer.gnome.org/glib/stable/glib-String-Utility-Functions.html#g-strcmp0](https://developer.gnome.org/glib/stable/glib-String-Utility-Functions.html#g-strcmp0)

if your program can crash when there is NULL pointer somewhere, than it is already programmed in a bad manner. And if you expect program to crash in some situations than you need to read CERT. 
[https://www.securecoding.cert.org/confluence/display/java/EXP01-J.+Do+not+use+a+null+in+a+case+where+an+object+is+required](https://www.securecoding.cert.org/confluence/display/java/EXP01-J.+Do+not+use+a+null+in+a+case+where+an+object+is+required)
[https://www.securecoding.cert.org/confluence/display/c/EXP34-C.+Do+not+dereference+null+pointers](https://www.securecoding.cert.org/confluence/display/c/EXP34-C.+Do+not+dereference+null+pointers)
[https://www.securecoding.cert.org/confluence/display/c/STR03-C.+Do+not+inadvertently+truncate+a+string](https://www.securecoding.cert.org/confluence/display/c/STR03-C.+Do+not+inadvertently+truncate+a+string)

And what is called "defensive programming" is a very good pattern. It decreases the number of vulnerabilities in programs.

And some other example I found from google tests:
[https://github.com/google/googletest/blob/master/googletest/samples/sample2_unittest.cc](https://github.com/google/googletest/blob/master/googletest/samples/sample2_unittest.cc)

As you can see this is the feature which is very useful.

**Comment by 9fcc** [comment]
In case of printf, - yes, it's bad practice. But only because POSIX standard says anything about NULL pointer passed to printf as its argument. So those who programmed printf functions made their decision. And I agree with them. But the very bad thing they've done is that they wrote nothing about NULL in man pages.

**Comment by chrisnc** [comment]
> We talk about tests. You know what values could be passed to your function. So you can add assert to check for NULL before comparing strings if it is erroneous for you.

You're proposing that everyone who expects the previous behavior add code to their tests when further up the thread you argued for your change to avoid writing some code in your own tests to account for what `NULL` means in your project.

I wonder, also, whether your project allows the empty string as a valid username, distinct from `NULL`. No software I've used has had this property that I'm aware of, and I wonder if you might simplify all your code by representing the absence of a username by an empty string, rather than changing a testing library to match your application's `NULL` semantics.

I think everyone in this thread understands the issues involved with `NULL` and the things you've linked. This is a matter of the principle of least surprise in interface design, and simply put, `ck_assert_str_eq(NULL, NULL)` succeeding is very surprising (especially if ge and le do not, violating their algebraic properties).

It's possible to write C such that not only are null pointers not used as sentinels or "empty" values, but are disallowed entirely (imagine no dynamic allocation, and all pointers derived from `&` or array to pointer decay). Indeed, much of the standard library is written in a way where null checking is the caller's responsibility, giving developers the freedom to either design their programs such that `NULL` can never occur, or simply to check for `NULL` themselves. An API that looks like a standard library API, except that it defines a different behavior for `NULL`, will confuse and cause subtle bugs.

Colin's blog post argues that many things which are described as "defensive programming" are actually just ways of preventing a process from exiting in the face of errors, when often an abort/assertion failure is the best outcome. So yes, be defensive and check for errors, but don't then proceed like everything is fine after an error is encountered. Failing an internal assert on any `NULL` argument to a str or mem assertion is consistent with this.

**Comment by 9fcc** [comment]
> You're proposing that everyone who expects the previous behavior add code to their tests when further up the thread you argued that this change was necessary to avoid writing some code in your own tests to account for what NULL means in your project. Not every application uses NULL in the way that you are.

Well, if they can have `NULL` as output value and it's an error, then yes, I think they must add a check in their tests for `NULL`. Such a way debugging would be much easier than debugging through Valgrind. But if not, they would have the same behaviour of the tests as in previous version. Why? Because in most situations people use one of the arguments as a right one (set in the test). So it could not be NULL because in such a case their tests wouldn't work. So their tests would fail and they would know why. Why do you think it's a bad idea?

> 
I wonder, also, whether your project allows the empty string as a valid username, distinct from NULL.

I think you understand that there only two ways to resolve this situation. The first is NULL. And the second is boolean flag (but why if NULL could be used?).

**Comment by 9fcc** [comment]
May it be that I don't understand what you're trying to say... Do you mean that we need to replace passing `NULL` directly to `printf` with string "(null)"? The result would be the same so yes I can do so.

**Comment by chrisnc** [comment]
> Well, if they can have NULL as output value and it's an error, then yes, I think they must add a check in their tests for NULL. Such a way debugging would be much easier than debugging through Valgrind. But if not, they would have the same behaviour of the tests as in previous version. Why? Because in most situations people use one of the arguments as a right one (set in the test). So it could not be NULL because in such a case their tests wouldn't work. So their tests would fail and they would know why. Why do you think it's a bad idea?

This is resolved by adding assertions against `NULL` to all of these macros as was suggested earlier. My only issue is with allowing an comparison assertion to succeed with two `NULL` arguments, because this would cause its behavior to differ from `strcmp` and `memcmp`, which do not allow `NULL`.

> I think you understand that there only two ways to resolve this situation. The first is NULL. And the second is boolean flag (but why if NULL could be used instead?).

Using an empty string as a sentinel value is very common. Indeed, in many languages an empty string is falsey in boolean contexts (e.g., Python, JavaScript). And in bash these situations are (almost) identical; a variable that has not been set expands to the empty string. Are you saying that your application allows the username field to be empty? Do you have tests to this effect?

**Comment by 9fcc** [comment]
> My only issue is with allowing a comparison assertion to succeed with two NULL arguments

Those situations description added in the documentation. So I don't see any problems. Old code would work as expected (except comparison of two strings that were got from tested object, but here a programmer must add `NULL` check if such a value could be). API could be changed because it is still zero (development) version. The same version policy has the project I work for even it is commercial. Users know that something can change in future versions.

As google tests use `NULL` comparison too I think it's a common practice.

> Using an empty string as a sentinel value is very common 

It is a decision that makes program less universal (more checks, more code, more work, code would be expensive), a little slower (you need to read memory location by pointer to determine null character instead of comparing pointer) and need much more maintenance (for example if strings are always dynamically allocated you can't use constant values as empty string, you'll need to make malloc for every empty string so it can be easily freed in future). So I won't ever use such algorithms. In any case it's bad example.

**Comment by 9fcc** [comment]
Now I have some time to show some example.

```c
char *get_some_original_str(void)
{
    char *s = strdup(original_str);
    // Here is a common error. No check for NULL.
    return s;
}

char *get_some_processed_str(void)
{
    char *s = strdup(processed_str);
    // Here is a common error. No check for NULL.
    return s;
}

START_TEST(test_strings_equal)
{
  record_test_name(tcase_name());

  char *s1 = get_some_original_str(); // Cannot return NULL.
  char *s2 = get_some_processed_str(); // Cannot return NULL, too.

  ck_assert_ptr_ne(s1, NULL); // if you don't have this check, it is your own mistake.
  ck_assert_ptr_ne(s2, NULL); // if you don't have this check, it is your own mistake, the second time.

  ck_assert_str_eq(s1, s2); // If code written carefully, there is no way NULL will get into this check.

  free(s1);
  free(s2);
}
END_TEST
```

This example's functions violate CERT rule "EXP34-C. Do not dereference null pointers":
[https://www.securecoding.cert.org/confluence/display/c/EXP34-C.+Do+not+dereference+null+pointers](https://www.securecoding.cert.org/confluence/display/c/EXP34-C.+Do+not+dereference+null+pointers)
If the test wouldn't have check for NULL, then it would violate the rule, too.

That is what I tried to explain.

**Comment by brarcher** [comment]
Thanks for the discussion, @9fcc and @chrisnc. To date the str macros did not document if NULL was allowed, but functionally if NULL was provided the test program would segfault. From this discussion it would seem that it may be best to not change the semantics of the existing strings macros to preserve backwards compatibility. Perhaps instead the existing macros could be updated to properly fail if a NULL is passed such that the failure is more intentional.

There may also be times when comparing strings which may be NULL is valuable. May I propose that for these use cases additional macros be defined. For example, using the example above, perhaps the following:
```
ck_assert_str_eq_allownull(s1, s2);
```
(The name could be better, but...) With one or more new macros the semantics of allowing NULL on string comparisons could be added without changing backwards compatibility. Is this acceptable?

**Comment by chrisnc** [comment]
@brarcher that all sounds fine to me. Adding an explicit assert to the existing macros to fail on `NULL` would be great. Adding new APIs that have explicitly different behavior for `NULL`  (communicated in the name itself, rather than just docs) is fine too. I agree that they can be useful for applications which use `NULL` in certain ways, but I think the defaults should follow the conventions used in the C standard library to minimize surprises.

**Comment by 9fcc** [comment]
@brarcher and @chrisnc. Well, I still think that it would be just much of duplicating macros but ok. I think `ck_assert_pstr_eq(s1, s2);`would be nice. `P`means pointer. Easy to remember and clear for understanding.

**Comment by 9fcc** [comment]
I've made changes so now it always fails when compared against NULL. pstr macros would be in another merge request..
